packages/eslint-plugin/src/types/rules.ts

```diff
@@ -4,30 +4,16 @@ type CodePath = Rule.CodePath;
 type CodePathSegment = Rule.CodePathSegment;
 
 interface TSRuleListener {
-  onCodePathStart?(codePath: CodePath, node: TSESTree.Node): void;
-
-  onCodePathEnd?(codePath: CodePath, node: TSESTree.Node): void;
-
-  onCodePathSegmentStart?(segment: CodePathSegment, node: TSESTree.Node): void;
-
-  onCodePathSegmentEnd?(segment: CodePathSegment, node: TSESTree.Node): void;
-
-  onCodePathSegmentLoop?(
-    fromSegment: CodePathSegment,
-    toSegment: CodePathSegment,
-    node: TSESTree.Node,
-  ): void;
-
   [key: string]:
+    | (() => void)
+    | ((node: TSESTree.Identifier) => void)
     | ((node: TSESTree.ImportDeclaration) => void)
     | ((node: TSESTree.CallExpression) => void)
     | ((codePath: CodePath, node: TSESTree.Node) => void)
     | ((segment: CodePathSegment, node: TSESTree.Node) => void)
-    | ((
-        fromSegment: CodePathSegment,
-        toSegment: CodePathSegment,
-        node: TSESTree.Node,
-      ) => void)
+    | ((node: TSESTree.FunctionDeclaration) => void)
+    | ((node: TSESTree.FunctionExpression) => void)
+    | ((node: TSESTree.ArrowFunctionExpression) => void)
```

Comment:
Can we just make this `(node: TSESTree.Node) => void`?

Reply:
We tried together. The exact error is 

`Type '(node: CallExpression) => void' is not assignable to type '(() => void) | ((node: Node) => void) | undefined'.` and so on.

Happy to work on this at some point though.